pkg/tools/opautils/opainput.go

```diff
@@ -0,0 +1,125 @@
+//    2) "auth_info" with sub keys:
+//          i) "certificate" -- is a pem encoded x509cert (usage: input.auth_info.certificate)
+//			ii) "spiffe_id" -- is a spiffeID from SVIDX509Certificate (usage: input.auth_info.spiffe_id)
+//    3) "operation" -- one of request/close (usage: input.operation)
+//    4) "role" -- one of client/endpoint (usage: input.role)
+
+// An example of using OPA input for the case of token signature verification:
+//
+//		package test
+//
+//		default allow = false
+//
+//		allow {
+//			token := input.connection.path.path_segments[0].token
+//			cert := input.auth_info.certificate
+//			io.jwt.verify_es256(token, cert) # signature verification
+//		}
+//
+
+// Package opautils provides of utilities for using OPA
+package opautils
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/spiffeutils"
+
+	"github.com/pkg/errors"
+
+	"google.golang.org/grpc/credentials"
+)
+
+type role string
+type operation string
+
+const (
+	// Client is a client role
+	Client role = "client"
+
+	// Endpoint is an endpoint role
+	Endpoint role = "endpoint"
+
+	// Request is a request operation
+	Request operation = "request"
+
+	// Close is a close operation
+	Close operation = "close"
+)
+
+// PreparedOpaInput - returns a prepared input for using in OPA
+func PreparedOpaInput(connection *networkservice.Connection, authInfo credentials.AuthInfo, operation operation, role role) (map[string]interface{}, error) {
+	connectionAsMap, err := convertConnectionToMap(connection)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Cannot convert connection %v to map", connection)
+	}
+
+	cert := parseX509Cert(authInfo)
+
+	spiffeID, err := spiffeutils.SpiffeIDFromX509(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	pemcert := pemEncodingX509Cert(cert)
+
+	rv := map[string]interface{}{
+		"connection": connectionAsMap,
+		"auth_info": map[string]interface{}{
+			"certificate": pemcert,
+			"spiffe_id":   spiffeID,
+		},
+		"operation": operation,
+		"role":      role,
+	}
+	return rv, nil
+}
+
+func pemEncodingX509Cert(cert *x509.Certificate) string {
+	certpem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})
+	return string(certpem)
+}
+
+func parseX509Cert(authInfo credentials.AuthInfo) *x509.Certificate {
+	var peerCert *x509.Certificate
+	if tlsInfo, ok := authInfo.(credentials.TLSInfo); ok {
+		if len(tlsInfo.State.PeerCertificates) > 0 {
+			peerCert = tlsInfo.State.PeerCertificates[0]
+		}
+	}
+	return peerCert
+}
+
+func convertConnectionToMap(connection *networkservice.Connection) (map[string]interface{}, error) {
+	jsonConn, err := json.Marshal(connection)
+	if err != nil {
+		return nil, err
+	}
+	rv := make(map[string]interface{})
+	if err := json.Unmarshal(jsonConn, &rv); err != nil {
```

Comment:
I am really curious why the unmarshal here :)

Reply:
I thought, that we need to convert all policy input data to a map[string]interface{}

Reply:
Ah, I see :)